samza-core/src/main/java/org/apache/samza/zk/ZkDistributedLock.java

```diff
@@ -0,0 +1,108 @@
+ */
+public class ZkDistributedLock implements DistributedLock {
+
+  public static final Logger LOG = LoggerFactory.getLogger(ZkDistributedLock.class);
+  private final static String LOCK_PATH = "distributed_lock";
+  private final ZkUtils zkUtils;
+  private final String lockPath;
+  private final String participantId;
+  private final ZkKeyBuilder keyBuilder;
+  private final Random random = new Random();
+  private String nodePath = null;
+
+  public ZkDistributedLock(String participantId, ZkUtils zkUtils, ZkConfig zkConfig) {
+    this.zkUtils = zkUtils;
+    this.participantId = participantId;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    lockPath = String.format("%s/%s", keyBuilder.getRootPath(), LOCK_PATH);
+    zkUtils.makeSurePersistentPathsExists(new String[] {lockPath});
+  }
+
+  /**
+   * Tries to acquire a lock in order to create intermediate streams. On failure to acquire lock, it keeps trying until the lock times out.
+   * Creates a sequential ephemeral node to acquire the lock. If the path of this node has the lowest sequence number, the processor has acquired the lock.
+   * @param timeout Duration of lock acquiring timeout.
+   * @param unit Unit of the timeout defined above.
+   * @return true if lock is acquired successfully, false if it times out.
+   */
+  @Override
+  public boolean lock(long timeout, TimeUnit unit) {
+    nodePath = zkUtils.getZkClient().createEphemeralSequential(lockPath + "/", participantId);
+
+    //Start timer for timeout
+    long startTime = System.currentTimeMillis();
+    long lockTimeout = TimeUnit.MILLISECONDS.convert(timeout, unit);
+
+    while (true) {
+      List<String> children = zkUtils.getZkClient().getChildren(lockPath);
+      int index = children.indexOf(ZkKeyBuilder.parseIdFromPath(nodePath));
+
+      if (children.size() == 0 || index == -1) {
+        throw new SamzaException("Looks like we are no longer connected to Zk. Need to reconnect!");
+      }
+      // Acquires lock when the node has the lowest sequence number and returns.
+      if (index == 0) {
+        LOG.info("Acquired lock for participant id: {}", participantId);
+        return true;
+      } else {
+        // Keep trying to acquire the lock
+        long currentTime = System.currentTimeMillis();
+        if ((currentTime - startTime) >= lockTimeout) {
+          return false;
+        }
+        try {
+          Thread.sleep(random.nextInt(1000));
+        } catch (InterruptedException e) {
+          Thread.interrupted();
+        }
+        LOG.info("Trying to acquire lock again...");
+      }
+    }
+  }
+
+  /**
+   * Unlocks, by deleting the ephemeral sequential node created to acquire the lock.
+   */
+  @Override
+  public void unlock() {
+    if (nodePath != null) {
+      zkUtils.getZkClient().delete(nodePath);
+      nodePath = null;
+      LOG.info("Ephemeral lock node deleted. Unlocked!");
+    } else {
+      LOG.error("Ephemeral lock node you want to delete doesn't exist");
```

Comment:
Perhaps change this `warn` since it doesn't actually indicate an error condition.